File: flutter_sdk/lib/src/domain/chat_message/chat_message.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'package:equatable/equatable.dart';

enum MessageRole {
  user('USER'),
  assistant('ASSISTANT');

  final String role;

  const MessageRole(this.role);
}

enum MessageType {
  text('text'),
  image('image'),
  voice('voice');

  final String type;
  const MessageType(this.type);
}

enum MessageStatus {
  delivered('DELIVERED'),
  read('READ'),
  error('ERROR'),
  sent('SENT'),
  inProgress('IN_PROGRESS');

  final String status;
  const MessageStatus(this.status);
}

// A class that represents a chat message in the chat
class ChatMessage extends Equatable {
  final int? id;
  final MessageRole role;
  final String content;
  final MessageType type;
  final MessageStatus status;
  final DateTime timestamp;

  // Audio messages fields
  final String? fileName;
  final List<double>? amplitudes;
  // Audio duration in ms
  final int? duration;

  const ChatMessage({
    this.id,
    required this.role,
    required this.type,
    required this.timestamp,
    this.content = '',
    this.status = MessageStatus.inProgress,
    this.fileName,
    this.amplitudes,
    this.duration,
  });

  // Creates a copy of a chat message
  ChatMessage copyWith({
    int? id,
    MessageRole? role,
    String? content,
    MessageType? type,
    MessageStatus? status,
    String? fileName,
    List<double>? amplitudes,
    int? duration,
    DateTime? timestamp,
  }) {
    return ChatMessage(
      id: id ?? this.id,
      role: role ?? this.role,
      content: content ?? this.content,
      type: type ?? this.type,
      status: status ?? this.status,
      fileName: fileName ?? this.fileName,
      amplitudes: amplitudes ?? this.amplitudes,
      duration: duration ?? this.duration,
      timestamp: timestamp ?? this.timestamp,
    );
  }

  // Equatable props array
  @override
  List<Object?> get props => [
    id,
    role,
    content,
    type,
    status,
    fileName,
    amplitudes,
    duration,
    timestamp,
  ];
}

File: typescript-sdk/packages/core/src/domain/chat-message.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import type { Product } from './product.js';

export enum MessageRole {
  User = 'USER',
  Assistant = 'AGENT',
}

export enum MessageType {
  Text = 'text',
  Image = 'image',
  Voice = 'voice',
  Product = 'product',
  ProductCarousel = 'productCarousel',
  Promotion = 'promotion',
  QuickReply = 'quickReply',
  Unknown = 'unknown',
}

export enum MessageStatus {
  Delivered = 'DELIVERED',
  Read = 'READ',
  Error = 'ERROR',
  Sent = 'SENT',
  InProgress = 'IN_PROGRESS',
}

export interface ChatMessage {
  /** Local DB auto-increment id (undefined before first save) */
  id?: number;
  /** Workflow-interpreter message id */
  wiId?: string;
  role: MessageRole;
  content: string;
  type: MessageType;
  status: MessageStatus;
  /** Unix epoch milliseconds */
  timestamp: number;
  /** File path or URL for image/audio */
  fileName?: string;
  /** Audio waveform amplitudes (preview) */
  amplitudes?: number[];
  /** Audio duration in milliseconds */
  duration?: number;
  products: Product[];
  /** UI-only: expanded carousel state (not persisted) */
  expand: boolean;
  quickReplies: string[];
}

// ── Named constructors ──────────────────────────────────────────────────────

export function chatMessageText(opts: {
  id?: number;
  wiId?: string;
  role: MessageRole;
  timestamp: number;
  status?: MessageStatus;
  content: string;
  quickReplies?: string[];
}): ChatMessage {
  return {
    id: opts.id,
    wiId: opts.wiId,
    role: opts.role,
    content: opts.content,
    type: MessageType.Text,
    status: opts.status ?? MessageStatus.InProgress,
    timestamp: opts.timestamp,
    products: [],
    expand: false,
    quickReplies: opts.quickReplies ?? [],
  };
}

export function chatMessageVoice(opts: {
  id?: number;
  wiId?: string;
  role: MessageRole;
  timestamp: number;
  status?: MessageStatus;
  fileName: string;
  amplitudes: number[];
  duration: number;
  quickReplies?: string[];
}): ChatMessage {
  return {
    id: opts.id,
    wiId: opts.wiId,
    role: opts.role,
    content: '',
    type: MessageType.Voice,
    status: opts.status ?? MessageStatus.InProgress,
    timestamp: opts.timestamp,
    fileName: opts.fileName,
    amplitudes: opts.amplitudes,
    duration: opts.duration,
    products: [],
    expand: false,
    quickReplies: opts.quickReplies ?? [],
  };
}

export function chatMessageImage(opts: {
  id?: number;
  wiId?: string;
  role: MessageRole;
  timestamp: number;
  fileName: string;
  status?: MessageStatus;
  content?: string;
  quickReplies?: string[];
}): ChatMessage {
  return {
    id: opts.id,
    wiId: opts.wiId,
    role: opts.role,
    content: opts.content ?? '',
    type: MessageType.Image,
    status: opts.status ?? MessageStatus.InProgress,
    timestamp: opts.timestamp,
    fileName: opts.fileName,
    products: [],
    expand: false,
    quickReplies: opts.quickReplies ?? [],
  };
}

export function chatMessageProduct(opts: {
  id?: number;
  wiId?: string;
  role: MessageRole;
  timestamp: number;
  status?: MessageStatus;
  products?: Product[];
  expand?: boolean;
  quickReplies?: string[];
}): ChatMessage {
  return {
    id: opts.id,
    wiId: opts.wiId,
    role: opts.role,
    content: '',
    type: MessageType.Product,
    status: opts.status ?? MessageStatus.InProgress,
    timestamp: opts.timestamp,
    fileName: '',
    products: opts.products ?? [],
    expand: opts.expand ?? false,
    quickReplies: opts.quickReplies ?? [],
  };
}

export function chatMessageCarousel(opts: {
  id?: number;
  wiId?: string;
  role: MessageRole;
  timestamp: number;
  status?: MessageStatus;
  products?: Product[];
  expand?: boolean;
  quickReplies?: string[];
}): ChatMessage {
  return {
    id: opts.id,
    wiId: opts.wiId,
    role: opts.role,
    content: '',
    type: MessageType.ProductCarousel,
    status: opts.status ?? MessageStatus.InProgress,
    timestamp: opts.timestamp,
    fileName: '',
    products: opts.products ?? [],
    expand: opts.expand ?? false,
    quickReplies: opts.quickReplies ?? [],
  };
}

export function chatMessageCopyWith(msg: ChatMessage, patch: Partial<ChatMessage>): ChatMessage {
  return { ...msg, ...patch };
}
